librad/src/keys.rs

```diff
@@ -15,5 +15,274 @@
 // You should have received a copy of the GNU General Public License
 // along with this program. If not, see <https://www.gnu.org/licenses/>.
 
-pub mod device;
-pub mod pgp;
+use std::{
+    fmt::{self, Display},
+    ops::Deref,
+};
+
+use bit_vec::BitVec;
+use secstr::SecStr;
+use serde::{Deserialize, Serialize};
+use sodiumoxide::crypto::sign::ed25519;
+
+use keystore::SecretKeyExt;
+
+pub use ed25519::PUBLICKEYBYTES;
+
+/// A device-specific signing key
+#[derive(Clone, Eq, PartialEq)]
+pub struct SecretKey(ed25519::SecretKey);
+
+/// The public part of a `Key``
+#[derive(Clone, Debug, Eq, PartialEq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
+pub struct PublicKey(ed25519::PublicKey);
+
+/// A signature produced by `Key::sign`
+#[derive(Clone, Debug, Eq, PartialEq, PartialOrd, Ord, Serialize, Deserialize)]
+pub struct Signature(ed25519::Signature);
+
+// Key
+
+#[allow(clippy::new_without_default)]
+impl SecretKey {
+    pub fn new() -> Self {
+        let (_, sk) = ed25519::gen_keypair();
+        Self(sk)
+    }
+
+    #[cfg(test)]
+    pub fn from_seed(seed: &ed25519::Seed) -> Self {
+        let (_, sk) = ed25519::keypair_from_seed(seed);
+        Self(sk)
+    }
+
+    pub(crate) fn from_secret(sk: ed25519::SecretKey) -> Self {
+        Self(sk)
+    }
+
+    pub fn public(&self) -> PublicKey {
+        PublicKey(self.0.public_key())
+    }
+
+    pub fn sign(&self, data: &[u8]) -> Signature {
+        Signature(ed25519::sign_detached(data, &self.0))
+    }
+
+    const PKCS_ED25519_OID: &'static [u64] = &[1, 3, 101, 112];
```

Comment:
What's this? :thinking:

Reply:
It's a magical value encoding the signature algorithm in a PKCS-encoded certificate. I cannot document this, I'm sorry, it is only accessible to initiated members of the free masons guild "Templars of Londres Shacklewell" (TLS).